extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -68,10 +75,16 @@ public HttpAuthenticator(IdentityProviderManager identityProviderManager,
                     break;
                 }
             }
-            // Add mechanism if there is a provider with matching credential type
-            // If the mechanism has no credential types, just add it anyway
-            if (found || mechanism.getCredentialTypes().isEmpty()) {
+            // Add mechanism if there is a provider with matching credential type or fail if there isn't
+            // We do not fail for the basic auth mechanism as it is added by default, not required by users
+            if (found) {
                 mechanisms.add(mechanism);
+            } else if (!BasicAuthenticationMechanism.class.equals(mechanism.getClass())) {
```

Comment:
@michalvavrik I propose to do a warning for all cases just so that we indeed avoid breaking something, appreciate your concern, rather than have mechanism specific exclusions. Warning should be fine because if the basic authentication is enabled by default but no matching identity provider exists it is still not working. Or how about debug level message for Basic if the warning would be too noisy but do warn for every other mechanism ?

Reply:
This was a tough one, I think I found right solution but please fight back if you disagree.

1. It is absolutely fine and user friendly to fail with an exception if mechanism is enabled explicitly or added by user but required identity provider is not there. Therefore I thrown an exception there.
2. Basic auth mechanism is added very often, we might need to refactor that in the future, but as long as it is added by default we shouldn't fail as it would be super breaking change. And user didn't explicitly require that mechanism. Therefore it is ignored with debug logging.
3. If basic auth mechanism is explicitly enabled but no provider was found, we should fail a build as it is good information for user.

I only run tests in Vert.x HTTP and Elytron properties modules, so let's see what CI thinks about that as well.